src/App/IndexedName.cpp

```diff
@@ -0,0 +1,171 @@
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <cstdlib>
+# include <unordered_set>
+#endif
+
+#include <QHash>
+
+#include "IndexedName.h"
+
+using namespace Data;
+
+struct ByteArray
+{
+    explicit ByteArray(QByteArray other)
+        :bytes(std::move(other))
+    {}
+
+    ByteArray(const ByteArray& other) = default;
+
+    ByteArray(ByteArray&& other) noexcept
+        :bytes(std::move(other.bytes))
+    {}
+
+    ~ByteArray() = default;
+
+    void ensureUnshared() const
+    {
+        QByteArray copy;
+        copy.append(bytes.constData(), bytes.size());
+        bytes = copy;
+    }
+
+    bool operator==(const ByteArray& other) const {
+        return bytes == other.bytes;
+    }
+
+    ByteArray &operator=(const ByteArray & other) {
+        bytes.clear();
+        bytes.append(other.bytes.constData(), other.bytes.size());
+        return *this;
+    }
+
+    ByteArray &operator= (ByteArray&& other) noexcept
+    {
+        bytes = std::move(other.bytes);
+        return *this;
+    }
+
+    mutable QByteArray bytes;
+};
+
+struct ByteArrayHasher
+{
+    std::size_t operator()(const ByteArray& bytes) const
+    {
+        return qHash(bytes.bytes);
+    }
+
+    std::size_t operator()(const QByteArray& bytes) const
+    {
+        return qHash(bytes);
+    }
+};
+
+inline bool isInvalidChar(char test)
+{
+    return test != '_' && (std::isalpha(test) == 0);
+}
+
+// Get the integer suffix of name. Returns a tuple of (suffix, suffixPosition). Calling code
+// should check to ensure that suffixPosition is not equal to length (in which case there was no
+// suffix).
+std::pair<int,int> getIntegerSuffix(const char *name, int length)
```

Comment:
I guess `length` is the length of `name`. Maybe this should be documented to avoid to wrongly use this function.

Reply:
I renamed the variable and added clearer documentation in d4d9ce7c18176d113713c330882ca4a10b68d55a